File: internal/relay/bundle.go
package relay

import (
	"net/http"
	"strings"

	"github.com/yacchi/backlog-cli/internal/config"
)

// AuditAction constants for bundle
const (
	AuditActionRelayBundle = "relay_bundle"
)

// handleRelayBundle は自動更新用のバンドルを返す
func (s *Server) handleRelayBundle(w http.ResponseWriter, r *http.Request) {
	domain := strings.TrimSpace(r.PathValue("domain"))
	if domain == "" {
		http.Error(w, "domain is required", http.StatusBadRequest)
		return
	}

	clientIP := ""
	if ip := getClientIP(r); ip != nil {
		clientIP = ip.String()
	}

	tenant, ok := findTenantByAllowedDomain(s.cfg.Server().Tenants, domain)
	if !ok {
		s.auditLogger.Log(AuditEvent{
			Action:    AuditActionRelayBundle,
			Domain:    domain,
			ClientIP:  clientIP,
			UserAgent: r.UserAgent(),
			Result:    "error",
			Error:     "tenant not found",
		})
		http.Error(w, "tenant not found", http.StatusNotFound)
		return
	}

	relayURL := s.buildRelayURL(r)
	bundleData, err := config.CreatePortalBundle(tenant, domain, relayURL)
	if err != nil {
		s.auditLogger.Log(AuditEvent{
			Action:    AuditActionRelayBundle,
			Domain:    domain,
			ClientIP:  clientIP,
			UserAgent: r.UserAgent(),
			Result:    "error",
			Error:     err.Error(),
		})
		http.Error(w, "failed to create bundle", http.StatusInternalServerError)
		return
	}

	space, backlogDomain := splitDomain(domain)
	s.auditLogger.Log(AuditEvent{
		Action:    AuditActionRelayBundle,
		Space:     space,
		Domain:    backlogDomain,
		ClientIP:  clientIP,
		UserAgent: r.UserAgent(),
		Result:    "success",
	})

	filename := domain + ".backlog-cli.zip"
	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition", "attachment; filename=\""+filename+"\"")
	_, _ = w.Write(bundleData)
}

File: packages/relay-core/src/handlers/bundle.ts
/**
 * Bundle handlers for configuration bundle distribution (no auth).
 */

import { Hono } from "hono";
import type { RelayConfig, AuditLogger, TenantConfig } from "../config/types.js";
import { AuditActions, createAuditEvent } from "../middleware/audit.js";
import { extractRequestContext } from "../utils/request.js";

/**
 * Find tenant by name.
 */
function findTenant(
  tenants: TenantConfig[] | undefined,
  name: string
): TenantConfig | undefined {
  return tenants?.find(
    (t) => t.name.toLowerCase() === name.toLowerCase()
  );
}

/**
 * Build relay URL from config and request context.
 */
function buildRelayUrl(
  baseUrl: string | undefined,
  reqBaseUrl: string
): string {
  return baseUrl || reqBaseUrl;
}

/**
 * Create bundle handlers (no authentication required).
 */
export function createBundleHandlers(
  config: RelayConfig,
  auditLogger: AuditLogger,
  createBundle: (
    tenant: TenantConfig,
    name: string,
    relayUrl: string
  ) => Promise<Uint8Array>
): Hono {
  const app = new Hono();

  /**
   * GET /v1/relay/tenants/:name/bundle - Download configuration bundle.
   * This endpoint does not require authentication.
   */
  app.get("/v1/relay/tenants/:name/bundle", async (c) => {
    const reqCtx = extractRequestContext(c);
    const name = c.req.param("name")?.trim();

    if (!name) {
      return c.text("name is required", 400);
    }

    const tenant = findTenant(config.tenants, name);
    if (!tenant) {
      auditLogger.log(
        createAuditEvent({
          action: AuditActions.RELAY_BUNDLE,
          domain: name,
          clientIp: reqCtx.clientIp,
          userAgent: reqCtx.userAgent,
          result: "error",
          error: "tenant not found",
        })
      );
      return c.text("tenant not found", 404);
    }

    const relayUrl = buildRelayUrl(config.server.base_url, reqCtx.baseUrl);

    try {
      const bundleData = await createBundle(tenant, name, relayUrl);

      auditLogger.log(
        createAuditEvent({
          action: AuditActions.RELAY_BUNDLE,
          domain: name,
          clientIp: reqCtx.clientIp,
          userAgent: reqCtx.userAgent,
          result: "success",
        })
      );

      const filename = `${name}.backlog-cli.zip`;
      return new Response(bundleData, {
        headers: {
          "Content-Type": "application/zip",
          "Content-Disposition": `attachment; filename="${filename}"`,
          "Cache-Control": "no-store",
        },
      });
    } catch (err) {
      auditLogger.log(
        createAuditEvent({
          action: AuditActions.RELAY_BUNDLE,
          domain: name,
          clientIp: reqCtx.clientIp,
          userAgent: reqCtx.userAgent,
          result: "error",
          error: (err as Error).message,
        })
      );
      return c.text("failed to create bundle", 500);
    }
  });

  return app;
}
